cmd/ci-operator/main_test.go

```diff
@@ -61,3 +76,204 @@ func TestSanitizeMessage(t *testing.T) {
+	testArtifactDirectory := filepath.Join(tempDir, "test-artifact-directory")
+	if os.Mkdir(testArtifactDirectory, os.FileMode(0755)) != nil {
+		return fmt.Errorf("unable to create artifact directory under temporary directory")
+	}
+
+	if len(customMetadata) > 0 {
+		testJSON, err := json.MarshalIndent(customMetadata, "", "")
+		if err != nil {
+			return fmt.Errorf("error marshalling custom metadata: %v", err)
+		}
+		err = ioutil.WriteFile(filepath.Join(testArtifactDirectory, "custom-prow-metadata.json"), []byte(testJSON), os.FileMode(0644))
+		if err != nil {
+			return fmt.Errorf("unable to create custom metadata file: %v", err)
+		}
+	}
+
+	// Write a bunch of empty files that should be ignored
+	err = ioutil.WriteFile(filepath.Join(testArtifactDirectory, "a-ignore1.txt"), []byte(""), os.FileMode(0644))
+	err = ioutil.WriteFile(filepath.Join(testArtifactDirectory, "b-ignore1.txt"), []byte(""), os.FileMode(0644))
+	err = ioutil.WriteFile(filepath.Join(testArtifactDirectory, "d-ignore1.txt"), []byte(""), os.FileMode(0644))
+	err = ioutil.WriteFile(filepath.Join(testArtifactDirectory, "e-ignore1.txt"), []byte(""), os.FileMode(0644))
+	if err != nil {
+		return fmt.Errorf("one or more of the empty *ignore files failed to write: %v", err)
+	}
+
+	// Sleep for one second to guarantee we get a different modified time of the second write
+	time.Sleep(1 * time.Second)
+	err = o.writeMetadataJSON()
+	if err != nil {
+		return fmt.Errorf("error while writing metadata JSON: %v", err)
+	}
+
+	err = verifyMetadataFields(metadataFile, org, repo, extraOrg, extraRepo, prowJobID, namespace, customMetadata)
+	if err != nil {
+		return fmt.Errorf("error validating metadata: %v", err)
+	}
+
+	info, err = os.Stat(metadataFile)
+	if err != nil {
+		return fmt.Errorf("unable to stat metadata file second time: %v", err)
+	}
+
+	hasCustomMetadata := len(customMetadata) > 0
+	if hasCustomMetadata && !info.ModTime().After(originalModTime) {
+		return fmt.Errorf("file was not modified after the second function call when it should have due to the presence of custom metadata")
+	} else if !hasCustomMetadata && info.ModTime().After(originalModTime) {
+		return fmt.Errorf("file was modified after the second function call when it should not have due to the absence of custom metadata")
+	}
+
+	return nil
+}
+
+func createOptions(tempDir string, org string, repo string, extraOrg string, extraRepo string, prowJobID string, namespace string) *options {
+	return &options{
+		artifactDir: tempDir,
+		jobSpec: &api.JobSpec{
+			JobSpec: downwardapi.JobSpec{
+				Refs: &prowapi.Refs{
+					Org:  org,
+					Repo: repo,
+				},
+				ExtraRefs: []prowapi.Refs{
+					{
+						Org:  extraOrg,
+						Repo: extraRepo,
+					},
+				},
+				ProwJobID: prowJobID,
+			},
+		},
+		namespace: namespace,
+	}
+}
+
+func verifyMetadataFields(metadataFile string, org string, repo string, extraOrg string, extraRepo string, prowJobID string, namespace string, customMetadata map[string]string) error {
```

Comment:
Similarly, I am not a fan of custom validation methods either. We should try to present the literal data that we want to validate. Whether that's raw JSON strings we expect to exist on disk or literal `struct`s once they're loaded, we should validate by either `ioutil.ReadFile()` and `string(expected) == string(read)` or load, then unmarshal and use `reflect.DeepEqual` to validate expected structures. 

When presenting a diff to a user, use `diff.ObjectReflectDiff` or `StringDiff` for rich and useful diffs on complex objects.

Reply:
FYI, I elected to use `go-cmp/cmp.Diff` as `diff.ObjectReflectDiff` comments note that it's deprecated in favor of go-cmp's version.